### PR TITLE
add ed25519 algorithm

### PR DESCRIPTION
**Pull request by NoelBright**
Add ed25519 module to crypto. And it can be selected by SetAlg(algChoice string)
function. "P256R1" for ECDSA and "ED25519" for ed25519.
This Ed25519 will be used to sign Sigchain, Transaction, Header and Block to replace
ECDSA.

Signed-off-by: NoelBright <noel.n.bright@gmail.com>

### Proposed changes in this pull request
Explain the changes in this pull request in order to help the project maintainers understand the overall impact of it.

### Type (put an `x` where ever applicable)
- [ ] Bug fix: Link to the issue
- [ ] Feature (Non-breaking change)
- [ ] Feature (Breaking change)
- [ ] Documentation Improvement

### Checklist
Please put an `x` against the checkboxes. Write a small comment explaining if its `N/A` (not applicable)

- [ ] Read the [CONTRIBUTION guidelines](https://github.com/nknorg/nkn#contributing).
- [ ] All the tests are passing after the introduction of new changes.
- [ ] Added tests respective to the part of code I have written.
- [ ] Added proper documentation where ever applicable (in code and README.md).
- [ ] Code has been written according to [NKN-Golang-Style-Guide](https://github.com/nknorg/nkn/wiki/NKN-Golang-Style-Guide)

### Extra information
Any extra information related to this pull request.

### CODE REVIEW

**Comment by yilunzhang** [comment]
Are we planning to remove all other curve support and only leave Ed25519? If not then I suggest to change all `else` to `else if algoChoice == P256R1`, otherwise it's fine to use `else` now.

**Comment by NoelBright** [comment]
Yes, I will change the logic of selection of signature algorithm in next commit.

**Comment by yilunzhang** [comment]
A tiny correction: the name is Ed25519 instead of ED25519

**Comment by NoelBright** [comment]
done

File: util/config/config.go [hunk 1]
```diff
@@ -24,12 +24,14 @@ const (
 
 const (
 	MaxNumTxnPerBlock   = 4096
+	MaxBlockSize        = 1 * 1024 * 1024
```
Comment (yilunzhang) [inline util/config/config.go]: We should add a comment about the unit (bit/byte) or rename the variable to include the unit

File: chain/blockValidator.go [hunk 2]
```diff
@@ -46,6 +46,16 @@ func TransactionCheck(block *Block) error {
 	if block.Transactions == nil {
 		return errors.New("empty block")
 	}
+	numTxn := len(block.Transactions)
+	if numTxn > config.MaxNumTxnPerBlock {
+		return errors.New("block contains too many transactions")
+	}
+
+	blockSize := block.GetTxsSize()
```
Comment (yilunzhang) [inline chain/blockValidator.go]: Should we also add header size here?

File: chain/mining.go [hunk 3]
```diff
@@ -46,13 +48,31 @@ func (bm *BuiltinMining) BuildBlock(height uint32, chordID []byte, winningHash c
...
+			continue
+		}
+
+		totalTxsSize = totalTxsSize + txn.GetSize()
+		if totalTxsSize > config.MaxBlockSize {
+			break
+		}
+
+		txCount++
+		if txCount >= config.MaxNumTxnPerBlock {
```
Comment (yilunzhang) [inline chain/mining.go]: Should this be > instead of >=?

**Comment by NoelBright** [comment]
fixed.

**Pull request merged**